Adding Redaction and Compliance docs (#45373)
dotnet/docs: docs/core/extensions/data-redaction.md

---
title: Data redaction
description: Learn how to use .NET data redaction libraries to protect your application's sensitive data.
ms.date: 03/17/2025
---

# Data redaction

Redaction helps you sanitize or mask sensitive info in logs, error messages, or other outputs. This keeps you compliant with privacy rules and protects sensitive data. It's useful in apps that handle personal data, financial info, or other confidential info.

> ## Install the package

Reviewer: ```suggestion
## Install redaction package
```

From the command line:

```console
dotnet add package Microsoft.Extensions.Compliance.Redaction
```

Or directly in the C# project file:

```xml
<ItemGroup>
  <PackageReference Include="Microsoft.Extensions.Compliance.Redaction" Version="[CURRENTVERSION]" />
</ItemGroup>

> ```

Reviewer: Same recommendation as mentioned earlier, about tabs here.

> ## Redactors

Reviewer: ```suggestion
## Available redactors
```

> Redactors handle the actual redaction of sensitive data. They replace or mask sensitive information.

Reviewer: ```suggestion
Redactors are responsible for the act of redacting sensitive data. They redact, replace, or mask sensitive information. Consider the following available redactors provided by the library:
```

- The <xref:Microsoft.Extensions.Compliance.Redaction.ErasingRedactor> replaces any input with an empty string.
- The <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactor> uses `HMACSHA256` to encode data being redacted.

## Usage example

> 
> ### Register the services
> 
> Register the services using one of the `AddRedaction` methods:

Reviewer: ```suggestion

To use the built-in redactors, you have to register the required services. Register the services using one of the available `AddRedaction` methods as described in the following list:
```

```csharp
public static IServiceCollection AddRedaction(this IServiceCollection services)
public static IServiceCollection AddRedaction(this IServiceCollection services, Action<IRedactionBuilder> configure)

> ```

Reviewer: Remove the code section here, and replace it with the following unordered list:

```markdown
- <xref:Microsoft.Extensions.DependencyInjection.RedactionServiceCollectionExtensions.AddRedaction(Microsoft.Extensions.DependencyInjection.IServiceCollection)>: Registers an implementation of <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider> in the <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection>.
- <xref:Microsoft.Extensions.DependencyInjection.RedactionServiceCollectionExtensions.AddRedaction(Microsoft.Extensions.DependencyInjection.IServiceCollection,System.Action{Microsoft.Extensions.Compliance.Redaction.IRedactionBuilder})>: Registers an implementation of <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider> in the <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection> and configures available redactors with the given `configure` delegate.
```

### Configure a redactor

> Fetch redactors at runtime using an `IRedactorProvider`. You can implement your own provider and register it inside the `AddRedaction` call, or use the default provider. Configure redactors using these `IRedactionBuilder` methods:

Reviewer: ```suggestion
Fetch redactors at runtime using an <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider>. You can implement your own provider and register it inside the `AddRedaction` call, or use the default provider. Configure redactors using these <xref:Microsoft.Extensions.Compliance.Redaction.IRedactionBuilder> methods:
```

```csharp
// This will use the default redactor, which is the ErasingRedactor

> builder.Services.AddRedaction();

Reviewer: Where does this `builder` instance come from? If we're going to not show that code, we should at least mention when type this builder is-I assume it's an `IHostApplicationBuilder`.

Author: I updated this to be using a serviceCollection instead of builder instance, so the document would be consistent.

// Using the default redactor provider:
builder.Services.AddRedaction(redactionBuilder =>
{
    // Assign a redactor to use for a set of data classifications.
    redactionBuilder.SetRedactor<MyRedactor>(MyTaxonomy.Private, MyTaxonomy.Personal);
    // Assign a fallback redactor to use when processing classified data for which no specific redactor has been registered.
    // The `ErasingRedactor` is the default fallback redactor. If no redactor is configured for a data classification then the data will be erased.
    redactionBuilder.SetFallbackRedactor<MyFallbackRedactor>();
});

// Using a custom redactor provider:
builder.Services.AddSingleton<IRedactorProvider, MyRedactorProvider>();
```

Given this data classification in your code:

```csharp
using Microsoft.Extensions.Compliance.Classification;

public static class MyTaxonomy
{
    public static string Name => "MyTaxonomy";

>     public static DataClassification Private => new DataClassification(Name, nameof(Private));
>     public static DataClassification Public => new DataClassification(Name, nameof(Public));
>     public static DataClassification Personal => new DataClassification(Name, nameof(Personal));

Reviewer: ```suggestion
    public static DataClassification Private => new(Name, nameof(Private));
    public static DataClassification Public => new(Name, nameof(Public));
    public static DataClassification Personal => new(Name, nameof(Personal));
```

}
```

### Configure the HMAC redactor

Configure the HMAC redactor using these `IRedactionBuilder` methods:

```csharp

> var serviceCollection = new ServiceCollection();

Reviewer: I'm confused as to why we're manually instantiating a `ServiceCollection` in this example, but in the earlier example, we were using a `builder`. I'd recommend that we try to stay consistent with this.

Author: I updated the previous example to be instantiating a ServiceCollection like this example for consistency.

serviceCollection.AddRedaction(builder =>
{
    builder.SetHmacRedactor(
        options =>
        {
            options.KeyId = 1234567890;
            options.Key = Convert.ToBase64String("1234567890abcdefghijklmnopqrstuvwxyz");
        },

        // Any data tagged with Personal or Private attributes will be redacted by the Hmac redactor.
        MyTaxonomy.Personal, MyTaxonomy.Private,

        // "DataClassificationSet" lets you compose multiple data classifications:
        // For example, here the Hmac redactor will be used for data tagged
        // with BOTH Personal and Private (but not one without the other).
        new DataClassificationSet(MyTaxonomy.Personal, MyTaxonomy.Private));
});
```

Alternatively, configure it this way:

```csharp
var serviceCollection = new ServiceCollection();
serviceCollection.AddRedaction(builder =>
{
    builder.SetHmacRedactor(
        Configuration.GetSection("HmacRedactorOptions"), MyTaxonomy.Personal);
});
```

Include this section in your JSON config file:

```json
{
    "HmacRedactorOptions": {
        "KeyId": 1234567890,
        "Key": "1234567890abcdefghijklmnopqrstuvwxyz"
    }
}
```

> - The `HmacRedactorOptions` requires its `KeyId` and `Key` properties to be set.

Reviewer: ```suggestion
- The <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactorOptions> requires its <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactorOptions.Key?displayProperty=nameWithType> and <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactorOptions.KeyId?displayProperty=nameWithType> properties to be set.
```

- The `Key` should be in base 64 format and at least 44 characters long. Use a distinct key for each major deployment of a service. Keep the key material secret and rotate it regularly.
- The `KeyId` is appended to each redacted value to identify the key used to hash the data.
- Different key IDs mean the values are unrelated and can't be used for correlation.

> > [!Note]

Reviewer: ```suggestion
> [!NOTE]
```

> > The `HmacRedactor` is still experimental, so the above methods will show warning `EXTEXP0002` indicating it's not yet stable.

Reviewer: ```suggestion
> The <xref:Microsoft.Extensions.Compliance.Redaction.HmacRedactor> is still experimental, so the preceding methods will cause the `EXTEXP0002` warningm indicating it's not yet stable.
```

---
title: Data redaction in .NET
description: Learn how to use .NET data redaction libraries to protect your application's sensitive data.

> ms.date: 03/17/2025

Reviewer: ```suggestion
ms.date: 03/20/2025
```

Author: done

---

# Data redaction in .NET

Redaction helps you sanitize or mask sensitive information in logs, error messages, or other outputs. This keeps you compliant with privacy rules and protects sensitive data. It's useful in apps that handle personal data, financial information, or other confidential data points.

## Install redaction package

> 

Reviewer: ```suggestion

To get started, install the [📦 Microsoft.Extensions.Compliance.Redaction](https://www.nuget.org/packages/Microsoft.Extensions.Compliance.Redaction) NuGet package:

```

```csharp
// This will use the default redactor, which is the ErasingRedactor
var serviceCollection = new ServiceCollection();
serviceCollection.AddRedaction();

// Using the default redactor provider:
serviceCollection.AddRedaction(redactionBuilder =>
{
    // Assign a redactor to use for a set of data classifications.
    redactionBuilder.SetRedactor<MyRedactor>(MyTaxonomy.Private, MyTaxonomy.Personal);
    // Assign a fallback redactor to use when processing classified data for which no specific redactor has been registered.
    // The `ErasingRedactor` is the default fallback redactor. If no redactor is configured for a data classification then the data will be erased.
    redactionBuilder.SetFallbackRedactor<MyFallbackRedactor>();
});

// Using a custom redactor provider:
builder.Services.AddSingleton<IRedactorProvider, MyRedactorProvider>();
```

Given this data classification in your code:

```csharp
using Microsoft.Extensions.Compliance.Classification;

> public static class MyTaxonomy

Reviewer: ```suggestion
public static class MyTaxonomyClassifications
```

Author: done

{
    public static string Name => "MyTaxonomy";

    public static DataClassification Private => new(Name, nameof(Private));
    public static DataClassification Public => new(Name, nameof(Public));
    public static DataClassification Personal => new(Name, nameof(Personal));
}
```

### Configure the HMAC redactor

> Configure the HMAC redactor using these `IRedactionBuilder` methods:

Reviewer: ```suggestion
Configure the HMAC redactor using these <xref:Microsoft.Extensions.Compliance.Redaction.IRedactionBuilder> methods:
```

### Configure a custom redactor

To create a custom redactor, define a subclass that inherits from <xref:Microsoft.Extensions.Compliance.Redaction.Redactor>:

```csharp
public sealed class StarRedactor : Redactor

public class StarRedactor : Redactor
{
    private const string Stars = "****";

    public override int GetRedactedLength(ReadOnlySpan<char> input) => Stars.Length;

    public override int Redact(ReadOnlySpan<char> source, Span<char> destination)
    {
        Stars.CopyTo(destination);

        return Stars.Length;
    }
}
```

### Create a custom redactor provider

The <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider> interface provides instances of redactors based on data classification. To create a custom redactor provider, inherit from <xref:Microsoft.Extensions.Compliance.Redaction.IRedactorProvider> as shown in the following example:

```csharp
using Microsoft.Extensions.Compliance.Classification;
using Microsoft.Extensions.Compliance.Redaction;

public sealed class StarRedactorProvider : IRedactorProvider
{
    private static readonly StarRedactor _starRedactor = new();

>     public static StarRedactorProvider Instance { get; } = new();

Reviewer: Is the `Instance` member needed? Also, for the `GetRedactor` API when we implement a custom one like shown here, when should we consider the given `DataClassificationSet` - what's the used for?
